src/main/java/it/gov/pagopa/payhub/activities/service/treasury/TreasuryOpiParserService.java

```diff
@@ -36,13 +58,38 @@ public List<String> parseData(Path ingestionFlowFilePath) {
             try {
                 flussoGiornaleDiCassa14 = treasuryUnmarshallerService.unmarshalOpi14(ingestionFlowFile);
                 log.debug("file flussoGiornaleDiCassa with Id {} parsed successfully ", flussoGiornaleDiCassa14.getId());
+                versione = TreasuryValidatorService.V_14;
             } catch (Exception exception) {
                 log.info("file flussoGiornaleDiCassa parsing error with opi 1.4 format {} ", exception.getMessage());
                 throw new TreasuryOpiInvalidFileException("Cannot parse treasury Opi file " + ingestionFlowFile);
             }
         }
-        //TODO in task 1658 it will be implemented the element to be returned
-        return List.of();
+
+
+        assert versione != null;
+        if (!treasuryValidatorService.validatePageSize(flussoGiornaleDiCassa14, flussoGiornaleDiCassa161, zipFileSize, versione)) {
```

Comment:
instead of having a single method inside which you will write many if/else based on the version, let's just split and move the logics on a separate service, calling it directly on the try/catch above to build the common output of both processes.

You could have 2 Services:
1. TreasuryOpiv14HandlerService
2. TreasuryOpiv161HandlerService

These services will work the related DTOs, in order to avoid classes where you are actually using Object and cast, or again a if based on the version field...

Such services could extend a common base abstract class using generics to reference the version related type.
This service will implement the Strategy design pattern (BaseTreasuryOpiHandlerService) which will contain the common algorithm to process the file:
```
Map<TreasuryOpEnum, List<Pair<TreasuryDTO, FlussoTesoreriaPIIDTO>>> handle(T treasuryOpiData, IngestionFlowFileDTO ingestionFlowFileDTO, int totalNumberOfTreasuryOpiFiles) {
   validateTotalNumberOfPages(treasuryOpiData, totalNumberOfTreasuryOpiFiles);
}

private void validateTotalNumberOfPages(T treasuryOpiData, int totalNumberOfTreasuryOpiFiles) {
   if(getTotalNumberOfTreasuryOpiFiles(treasuryOpiData) != totalNumberOfTreasuryOpiFiles){
      throw new TreasuryOpiInvalidFileException(...); // if you want to log also the file name, add it as input to your methods
   }
...
}

protected abstract int getTotalNumberOfTreasuryOpiFiles(T treasuryOpiData);
```

Reply:
done refactor according to Strategy pattern